api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/handler/GaugeHandler.java

```diff
@@ -300,8 +307,34 @@ public void findGaugeData(
                 percentiles = new Percentiles(Collections.<Double>emptyList());
             }
 
-            metricsService.findGaugeStats(metricId, timeRange.getStart(), timeRange.getEnd(), buckets,
-                    percentiles.getPercentiles())
+            Observable<TimeRange> observableTimeRange = Observable.just(timeRange);
+
+            if (Boolean.TRUE.equals(fromEarliest)) {
+                metricsService.findMetric(metricId).first().map((metric) -> {
+                    long now = System.currentTimeMillis();
+                    long earliest = now - metric.getDataRetention() * 24 * 60 * 60 * 1000L;
+                    return new TimeRange(earliest, now);
+                });
+            }
+
+            final Percentiles localPercentiles = percentiles;
+            observableTimeRange
```

Comment:
Note that these comments apply also to the corresponding changes in CounterHandler. First, The only thing I see being done when ```fromEarliest``` is true is that we filter out empty buckets. Don't we want to include the oldest data available which would be determined by the retention settings. It looks like this code is just using whatever the value is for the start time param. We are missing code to calculate the start time.

Secondly, this code can be simplified and written in a more functional way, e.g.,

```java
if (Boolean.TRUE.equals(fromEarliest)) {
    metricsService.findGaugeStats(metricId, timeRange.getStart(), timeRange.getEnd(), buckets,
            percentiles.getPercentiles())
            .filter(bucket -> !bucket.isEmpty())
            .map(ApiUtils::collectionToResponse)
            .subscribe(asyncResponse::resume, t -> asyncResponse.resume(ApiUtils.serverError(t)));
} else {
    metricsService.findGaugeStats(metricId, timeRange.getStart(), timeRange.getEnd(), buckets,
        percentiles.getPercentiles())
        .map(ApiUtils::collectionToResponse)
        .subscribe(asyncResponse::resume, t -> asyncResponse.resume(ApiUtils.serverError(t)));
}
```